packages/xrpc-server/src/stream/frames.ts

```diff
@@ -0,0 +1,122 @@
+import * as cborx from 'cbor-x'
+import * as uint8arrays from 'uint8arrays'
+import {
+  frameHeader,
+  FrameHeader,
+  FrameType,
+  InfoFrameHeader,
+  DataFrameHeader,
+  ErrorFrameHeader,
+  infoFrameBody,
+  InfoFrameBody,
+  ErrorFrameBody,
+  errorFrameBody,
+} from './types'
+
+export abstract class Frame {
+  header: FrameHeader
+  body: unknown
+  get op(): FrameType {
+    return this.header.op
+  }
+  toBytes(): Uint8Array {
+    return uint8arrays.concat([
+      cborx.encode(this.header),
+      cborx.encode(this.body),
+    ])
+  }
+  static fromBytes(bytes: Uint8Array) {
+    let i = 0
+    let header: unknown
+    let body: unknown = kUnset
+    cborx.decodeMultiple(bytes, (item) => {
+      if (i === 0) {
+        header = item
+      } else if (i === 1) {
+        body = item
+      } else {
+        throw new Error('Too many CBOR data items in frame')
+      }
+      i++
+    })
+    const parsedHeader = frameHeader.safeParse(header)
+    if (!parsedHeader.success) {
+      throw new Error(`Invalid frame header: ${parsedHeader.error.message}`)
+    }
+    if (body === kUnset) {
+      throw new Error('Missing frame body')
+    }
+    const frameOp = parsedHeader.data.op
+    if (frameOp === FrameType.Data) {
+      return new DataFrame({
+        type: parsedHeader.data.t,
+        body,
+      })
+    } else if (frameOp === FrameType.Info) {
+      const parsedBody = infoFrameBody.safeParse(body)
+      if (!parsedBody.success) {
+        throw new Error(`Invalid info frame body: ${parsedBody.error.message}`)
+      }
+      return new InfoFrame({
+        body: parsedBody.data,
+      })
+    } else if (frameOp === FrameType.Error) {
+      const parsedBody = errorFrameBody.safeParse(body)
+      if (!parsedBody.success) {
+        throw new Error(`Invalid error frame body: ${parsedBody.error.message}`)
+      }
+      return new ErrorFrame({
+        body: parsedBody.data,
+      })
+    } else {
+      const exhaustiveCheck: never = frameOp
+      throw new Error(`Unknown frame op: ${exhaustiveCheck}`)
+    }
+  }
+}
+
+export class DataFrame extends Frame {
+  header: DataFrameHeader
+  constructor(opts: { type?: number; body: unknown }) {
```

Comment:
is `type` optional here? 

I guess in the case of a stream with only one message type? 

part of me wants to stick it on their anyway just for consistency 😅

Reply:
The thought I was having is that it's only really useful for unions, and even in that case you don't absolutely need it because the fallback for any unmapped `$type` is to put it in the message body rather than in the header.  So it's more of an optimization.  But if we do want to include it, we'd need to come-up with a convention for how to map the single type since it doesn't have an nsid.  Would it just be `0`?  Does that maybe mess things up if you change it to a union later?

Reply:
One thing we could do is default it to `-1`, and say that mapped types are non-negative.